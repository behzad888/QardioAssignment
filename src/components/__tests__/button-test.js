//@flow
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Button from '../button/button';

const createEvent = (type, data) => {
  const event = document.createEvent('MouseEvent');
  event.initEvent(type, true, true);
  if (data != null) {
    Object.entries(data).forEach(([key, value]) => {
      event[key] = value;
    });
  }
  return event;
};

describe('while button tests running in a browser environment', () => {
  let container, wrapper;

  beforeEach(() => {
    jest.resetModules();

    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    ReactDOM.render(null, container);
    document.body.removeChild(container);
    container = null;
  });

  describe('color variants:', () => {
    it('should render red button', () => {
      wrapper = renderer
        .create(<Button label="This is a Button" color="Red" />)
        .toJSON();
      expect(wrapper).toMatchSnapshot();
    });
    it('should render gray button', () => {
      wrapper = renderer
        .create(<Button label="This is a Button" color="Gray" />)
        .toJSON();
      expect(wrapper).toMatchSnapshot();
    });

    it('should render blue button', () => {
      wrapper = renderer
        .create(<Button label="This is a Button" color="Blue" />)
        .toJSON();
      expect(wrapper).toMatchSnapshot();
    });

    it('should render white button', () => {
      wrapper = renderer
        .create(<Button label="This is a Button" color="White" />)
        .toJSON();
      expect(wrapper).toMatchSnapshot();
    });

    it('should render primary button', () => {
      wrapper = renderer
        .create(<Button label="This is a Button" color="Primary" />)
        .toJSON();
      expect(wrapper).toMatchSnapshot();
    });

    it('should render transparent button', () => {
      wrapper = renderer
        .create(<Button label="This is a Button" color="Transparent" />)
        .toJSON();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('link component with tag a', () => {
    it('should render as an anchor with https://www.getqardio.com', () => {
      wrapper = renderer
        .create(
          <Button
            label="This is a Link"
            component="a"
            href="https://www.getqardio.com"
          />,
        )
        .toJSON();
      expect(wrapper).toMatchSnapshot();
    });
    it('should render as an anchor with href="#"', () => {
      wrapper = renderer
        .create(<Button label="This is a Link" component="a" />)
        .toJSON();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('size variants:', () => {
    it('should render large button', () => {
      wrapper = renderer
        .create(<Button label="This is a Button" color="White" size="large" />)
        .toJSON();
      expect(wrapper).toMatchSnapshot();
    });
    it('should render medium button', () => {
      wrapper = renderer
        .create(<Button label="This is a Button" color="White" size="medium" />)
        .toJSON();
      expect(wrapper).toMatchSnapshot();
    });
    it('should render small button', () => {
      wrapper = renderer
        .create(<Button label="This is a Button" color="White" size="small" />)
        .toJSON();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('should render ghost and round', () => {
    it('should render ghost button', () => {
      wrapper = renderer
        .create(<Button label="This is a Button" ghost />)
        .toJSON();
      expect(wrapper).toMatchSnapshot();
    });

    it('should render round button', () => {
      wrapper = renderer
        .create(<Button label="This is a Button" round />)
        .toJSON();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('should render child', () => {
    it('should render with child node', () => {
      wrapper = renderer
        .create(
          <div>
            <Button label="This is a Button" />
            <Button>This is a Button</Button>
            <Button>
              <h2>This is a Button</h2>
            </Button>
          </div>,
        )
        .toJSON();
      expect(wrapper).toMatchSnapshot();
    });
    it('should render nested node', () => {
      wrapper = renderer
        .create(
          <div>
            <Button>
              <Button>
                <h2>This is a Button</h2>
              </Button>
            </Button>
          </div>,
        )
        .toJSON();
      expect(wrapper).toMatchSnapshot();
    });
  });
  it('should render busy with three dot', () => {
    wrapper = renderer
      .create(
        <div>
          <Button busy={true} label="This is a Button" />
          <Button busy={true}>
            <h3> This is a Button</h3>
          </Button>
        </div>,
      )
      .toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  describe('Event Responder', () => {
    let onClick, ref;

    beforeEach(() => {
      onClick = jest.fn();
      ref = React.createRef();
      const element = <Button onClick={onClick} ref={ref} />;
      ReactDOM.render(element, container);
    });

    it('ignores browser emulated events', () => {
      ref.current.dispatchEvent(createEvent('click'));
      expect(onClick).toHaveBeenCalledTimes(1);
    });
  });
});
