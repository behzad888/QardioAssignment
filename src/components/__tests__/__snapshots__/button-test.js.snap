// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`while button tests running in a browser environment color variants: should render blue button 1`] = `
<button
  className="btn large btn-Blue"
  onClick={[Function]}
>
  This is a Button
</button>
`;

exports[`while button tests running in a browser environment color variants: should render gray button 1`] = `
<button
  className="btn large btn-Gray"
  onClick={[Function]}
>
  This is a Button
</button>
`;

exports[`while button tests running in a browser environment color variants: should render primary button 1`] = `
<button
  className="btn large btn-Primary"
  onClick={[Function]}
>
  This is a Button
</button>
`;

exports[`while button tests running in a browser environment color variants: should render red button 1`] = `
<button
  className="btn large btn-Red"
  onClick={[Function]}
>
  This is a Button
</button>
`;

exports[`while button tests running in a browser environment color variants: should render transparent button 1`] = `
<button
  className="btn large btn-Transparent"
  onClick={[Function]}
>
  This is a Button
</button>
`;

exports[`while button tests running in a browser environment color variants: should render white button 1`] = `
<button
  className="btn large btn-White"
  onClick={[Function]}
>
  This is a Button
</button>
`;

exports[`while button tests running in a browser environment link component with tag a should render as an anchor with href="#" 1`] = `
<a
  className="btn large btn-White "
  href="#"
>
  This is a Link
</a>
`;

exports[`while button tests running in a browser environment link component with tag a should render as an anchor with https://www.getqardio.com 1`] = `
<a
  className="btn large btn-White "
  href="https://www.getqardio.com"
>
  This is a Link
</a>
`;

exports[`while button tests running in a browser environment should render busy with three dot 1`] = `
<div>
  <button
    className="btn large btn-White"
    onClick={[Function]}
  >
    ...
  </button>
  <button
    className="btn large btn-White"
    onClick={[Function]}
  >
    ...
  </button>
</div>
`;

exports[`while button tests running in a browser environment should render child should render nested node 1`] = `
<div>
  <button
    className="btn large btn-White"
    onClick={[Function]}
  >
    <button
      className="btn large btn-White"
      onClick={[Function]}
    >
      <h2>
        This is a Button
      </h2>
    </button>
  </button>
</div>
`;

exports[`while button tests running in a browser environment should render child should render with child node 1`] = `
<div>
  <button
    className="btn large btn-White"
    onClick={[Function]}
  >
    This is a Button
  </button>
  <button
    className="btn large btn-White"
    onClick={[Function]}
  >
    This is a Button
  </button>
  <button
    className="btn large btn-White"
    onClick={[Function]}
  >
    <h2>
      This is a Button
    </h2>
  </button>
</div>
`;

exports[`while button tests running in a browser environment should render ghost and round should render ghost button 1`] = `
<button
  className="btn large btn-White ghost"
  onClick={[Function]}
>
  This is a Button
</button>
`;

exports[`while button tests running in a browser environment should render ghost and round should render round button 1`] = `
<button
  className="btn large btn-White round"
  onClick={[Function]}
>
  This is a Button
</button>
`;

exports[`while button tests running in a browser environment size variants: should render large button 1`] = `
<button
  className="btn large btn-White"
  onClick={[Function]}
>
  This is a Button
</button>
`;

exports[`while button tests running in a browser environment size variants: should render medium button 1`] = `
<button
  className="btn medium btn-White"
  onClick={[Function]}
>
  This is a Button
</button>
`;

exports[`while button tests running in a browser environment size variants: should render small button 1`] = `
<button
  className="btn small btn-White"
  onClick={[Function]}
>
  This is a Button
</button>
`;
